### PR TITLE
Jenkinsfile: collect pods state and kubelet logs in caught exception

### PR DESCRIPTION
**Pull request by okartau**

### CODE REVIEW

**Comment by pohly** [comment]
Have you tested that this works?

File: Jenkinsfile [hunk 1]
```diff
@@ -525,6 +525,12 @@ void TestInVM(deviceMode, deploymentMode, distro, distroVersion, kubernetesVersi
                                          TEST_E2E_SKIP=\$(if [ \"${env.CHANGE_ID}\" ] && [ \"${env.CHANGE_ID}\" != null ]; then echo \\\\[Slow\\\\]; fi) \
                            ' \
            "
+    } catch (exc) {
+        echo "Handling exception, trying to get state and logs:"
+        sh '''
+            _work/${env.CLUSTER}/ssh.0 kubectl get pods --all-namespaces -o wide
+            _work/${env.CLUSTER}/ssh.0 sudo journalctl -u kubelet
```
Comment (pohly) [inline Jenkinsfile]: We want the kubelet log for all nodes in the cluster, not just the first one.

**Comment by okartau** [comment]
> Have you tested that this works?

good point! As we mostly get good runs in PR jobs, having change here in current form does not show much useful information.
I added same in another Jenkins where I try to debug this pipeline, and it appears variable ${env.CLUSTER} is not set inside catch-block, although this is not intuitive as it appears to be set in try-block. Seems we have to hard-code few values.

**Comment by okartau** [comment]
or is it so that triple-single-quote what I used, does not let me pass variable there, and I should use double quote.

**Comment by pohly** [comment]
> good point! As we mostly get good runs in PR jobs, having change here in current form does not show much useful information.

What I proposed in https://github.com/intel/pmem-csi/issues/529#issuecomment-586876944 was to artificially make the test run time out by replacing the actual test with a sleep. Combine that with shortening the timeout for the step and you get relatively fast feedback whether this works (still need to build and deploy the driver).

> or is it so that triple-single-quote what I used, does not let me pass variable there, and I should use double quote.

Yes, that's probably it.

File: Jenkinsfile [hunk 2]
```diff
@@ -525,6 +525,17 @@ void TestInVM(deviceMode, deploymentMode, distro, distroVersion, kubernetesVersi
                                          TEST_E2E_SKIP=\$(if [ \"${env.CHANGE_ID}\" ] && [ \"${env.CHANGE_ID}\" != null ]; then echo \\\\[Slow\\\\]; fi) \
                            ' \
            "
+    } catch (exc) {
+        echo "Handling exception, trying to get state and logs:"
+        sh "_work/${env.CLUSTER}/ssh.0 kubectl get pods --all-namespaces -o wide"
+        echo "====== KUBELET LOG MASTER ======"
+        sh "_work/${env.CLUSTER}/ssh.0 sudo journalctl -u kubelet"
+        echo "====== KUBELET LOG WORKER-1 ======"
```
Comment (pohly) [inline Jenkinsfile]: Please don't hard-code the number of nodes. You can iterate over `_work/${env.CLUSTER}/ssh.*` instead.

Comment (okartau) [reply Jenkinsfile]: removed hard coding, now runs using all ssh.* wrappers

**Comment by okartau** [comment]
Now there is version that gets activated on error (including timeout of stage), showing pods list and kubelet logs using all existing ssh.X.
Anything else we want to add?
The forcibly shortened timeout triggers abortion now. Just sleep is not enough as kube system is set up in first part of same stage.

**Comment by pohly** [comment]
Looks good to me (https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/job/pmem-csi/view/change-requests/job/PR-534/6/execution/node/113/log/?consoleFull). Let's remove the timeout change and merge this.

**Pull request merged**